Protect Local/Session storage ipc endpoints

darryl-apple opened the pull request:
#### 20c5072b3c465c03822d9ca2eb0cbfbb34d03592
<pre>
Protect Local/Session storage ipc endpoints
<a href="https://bugs.webkit.org/show_bug.cgi?id=296743">https://bugs.webkit.org/show_bug.cgi?id=296743</a>
<a href="https://rdar.apple.com/157216538">rdar://157216538</a>

Reviewed by Sihui Liu.

Add a SessionStorageEnabled preference which can be used along with LocalStorageEnabled to protect the relevant IPC endpoints

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/inspector/agents/InspectorDOMStorageAgent.cpp:
(WebCore::InspectorDOMStorageAgent::findStorageArea):
* Source/WebCore/page/LocalDOMWindow.cpp:
* Source/WebCore/page/WindowLocalStorage.idl:
* Source/WebCore/page/WindowSessionStorage.idl:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::connectToStorageArea):
(WebKit::NetworkStorageManager::setItem):
(WebKit::NetworkStorageManager::removeItem):
(WebKit::NetworkStorageManager::clear):
(WebKit::NetworkStorageManager::isStorageTypeEnabled const):
(WebKit::NetworkStorageManager::isStorageAreaTypeEnabled const):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:

Canonical link: <a href="https://commits.webkit.org/298311@main">https://commits.webkit.org/298311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8fdfd8d9ccd0de47a2407120ede26a03378d637

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34788 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25272 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121190 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65716 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b7319918-0948-4d4f-8323-b39b9e1c89c4) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116964 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35435 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43350 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87456 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3f1ab84e-9ed7-4d26-a729-81e4bbff9fa4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118023 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103314 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67852 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27424 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21434 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64843 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107351 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97632 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124378 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113618 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42044 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31444 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96249 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42414 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96035 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24438 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41234 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19072 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38045 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41919 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47455 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137823 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41460 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36834 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44779 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43195 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/inspector/agents/InspectorDOMStorageAgent.cpp:
```diff
@@ -218,9 +219,21 @@ RefPtr<StorageArea> InspectorDOMStorageAgent::findStorageArea(Inspector::Protoco
+        if (!m_inspectedPage->settings().sessionStorageEnabled()) {
+            errorString = "Session storage is disabled"_s;
+            return nullptr;
+        }
```
It seems this check is unnecessary as you already make sure the storage will not be created in `LocalDOMWindow::sessionStorage()` when flag is disabled.

darryl-apple replied:
yep you are right, I've removed the unnecessary check

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1572,18 +1572,26 @@ void NetworkStorageManager::connectToStorageArea(IPC::Connection& connection, We
     switch (type) {
     case WebCore::StorageType::Local:
+        if (preferences && !preferences->localStorageEnabled)
+            return completionHandler(std::nullopt, HashMap<String, String> { }, StorageAreaBase::nextMessageIdentifier());
```
Can we use `MESSAGE_CHECK_COMPLETION` instead?

darryl-apple replied:
Yes that's seems sensible, I've updated these to use MESSAGE_CHECK_COMPLETION instead

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1572,18 +1572,26 @@ void NetworkStorageManager::connectToStorageArea(IPC::Connection& connection, We
         resultIdentifier = originStorageManager->localStorageManager(*m_storageAreaRegistry).connectToLocalStorageArea(connectionIdentifier, sourceIdentifier, origin, m_queue.copyRef());
         break;
     case WebCore::StorageType::TransientLocal:
+        if (preferences && !preferences->localStorageEnabled)
```
`MESSAGE_CHECK_COMPLETION`?

darryl-apple replied:
done

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1572,18 +1572,26 @@ void NetworkStorageManager::connectToStorageArea(IPC::Connection& connection, We
         resultIdentifier = originStorageManager->localStorageManager(*m_storageAreaRegistry).connectToTransientLocalStorageArea(connectionIdentifier, sourceIdentifier, origin);
         break;
     case WebCore::StorageType::Session:
+        if (preferences && !preferences->sessionStorageEnabled)
```
`MESSAGE_CHECK_COMPLETION`?

darryl-apple replied:
done

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1665,6 +1673,13 @@ void NetworkStorageManager::setItem(IPC::Connection& connection, StorageAreaIden
 
     MESSAGE_CHECK_COMPLETION(isSiteAllowedForConnection(connection.uniqueID(), WebCore::RegistrableDomain { storageArea->origin().topOrigin }), connection, completionHandler(hasError, WTFMove(allItems)));
 
+    if (auto preferences = sharedPreferencesForWebProcess(connection)) {
```
`MESSAGE_CHECK_COMPLETION`?

darryl-apple replied:
done

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1688,6 +1703,13 @@ void NetworkStorageManager::removeItem(IPC::Connection& connection, StorageAreaI
 
     MESSAGE_CHECK_COMPLETION(isSiteAllowedForConnection(connection.uniqueID(), WebCore::RegistrableDomain { storageArea->origin().topOrigin }), connection, completionHandler(hasError, WTFMove(allItems)));
 
+    if (auto preferences = sharedPreferencesForWebProcess(connection)) {
```
`MESSAGE_CHECK_COMPLETION`?

darryl-apple replied:
done

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1707,6 +1729,13 @@ void NetworkStorageManager::clear(IPC::Connection& connection, StorageAreaIdenti
 
     MESSAGE_CHECK_COMPLETION(isSiteAllowedForConnection(connection.uniqueID(), WebCore::RegistrableDomain { storageArea->origin().topOrigin }), connection, completionHandler());
 
+    if (auto preferences = sharedPreferencesForWebProcess(connection)) {
```
`MESSAGE_CHECK_COMPLETION`?

darryl-apple replied:
done

darryl-apple committed:
Protect Local/Session storage ipc endpoints
https://bugs.webkit.org/show_bug.cgi?id=296743
rdar://157216538

Reviewed by Sihui Liu.

Add a SessionStorageEnabled preference which can be used along with LocalStorageEnabled to protect the relevant IPC endpoints

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/inspector/agents/InspectorDOMStorageAgent.cpp:
(WebCore::InspectorDOMStorageAgent::findStorageArea):
* Source/WebCore/page/LocalDOMWindow.cpp:
* Source/WebCore/page/WindowLocalStorage.idl:
* Source/WebCore/page/WindowSessionStorage.idl:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::connectToStorageArea):
(WebKit::NetworkStorageManager::setItem):
(WebKit::NetworkStorageManager::removeItem):
(WebKit::NetworkStorageManager::clear):
(WebKit::NetworkStorageManager::isStorageTypeEnabled const):
(WebKit::NetworkStorageManager::isStorageAreaTypeEnabled const):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:

Canonical link: https://commits.webkit.org/298311@main
